src/main/java/io/cdap/plugin/sap/ODataEntryRecordReader.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.plugin.sap;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.olingo.odata2.api.ep.entry.ODataEntry;
+import org.apache.olingo.odata2.api.ep.feed.ODataFeed;
+
+import java.io.IOException;
+import java.util.Iterator;
+
+/**
+ * RecordReader implementation, which reads OData v2 entries.
+ */
+public class ODataEntryRecordReader extends RecordReader<NullWritable, ODataEntry> {
+
+  private static final Gson gson = new GsonBuilder().create();
+
+  private Iterator<ODataEntry> iterator;
+  private ODataEntry value;
+
+  /**
+   * Initialize an iterator and config.
+   *
+   * @param inputSplit         specifies batch details
+   * @param taskAttemptContext task context
+   */
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext) {
+    Configuration conf = taskAttemptContext.getConfiguration();
+    String configJson = conf.get(ODataEntryInputFormatProvider.PROPERTY_CONFIG_JSON);
+    SapODataConfig config = gson.fromJson(configJson, SapODataConfig.class);
+
+    OData2Client oData2Client = new OData2Client(config.getUrl(), config.getUser(), config.getPassword());
+    ODataFeed feed = oData2Client.queryEntitySet(config.getResourcePath(), config.getQuery());
+    iterator = feed.getEntries().iterator();
+  }
+
+  @Override
+  public boolean nextKeyValue() {
+    if (!iterator.hasNext()) {
+      return false;
+    }
+    value = iterator.next();
+    return true;
+  }
+
+  @Override
+  public NullWritable getCurrentKey() {
+    return null;
+  }
+
+  @Override
+  public ODataEntry getCurrentValue() {
+    return value;
+  }
+
+  @Override
+  public float getProgress() {
+    // progress is unknown
+    return 0.0f;
+  }
+
+  @Override
+  public void close() throws IOException {
+  }
```

Comment:
I assume this means the client doesn't need to be closed?

Reply:
Yes. According to the [Olingo Client sample](https://olingo.apache.org/doc/odata2/tutorials/OlingoV2BasicClientSample.html#client-sample), there is no need to close the client.